delfin/drivers/ibm/storwize_svc/ssh_handler.py

```diff
@@ -0,0 +1,401 @@
+            msg = ("Failed to build alert model as some attributes missing "
+                   "in alert message:%s.") % (six.text_type(e))
+            raise exception.InvalidResults(msg)
+
+    def login(self):
+        try:
+            with self.ssh_pool.item() as ssh:
+                SSHHandler.do_exec('lssystem', ssh)
+        except Exception as e:
+            LOG.error("Failed to login ibm storwize_svc %s" %
+                      (six.text_type(e)))
+            raise e
+
+    @staticmethod
+    def do_exec(command_str, ssh):
+        """Execute command"""
+        try:
+            utils.check_ssh_injection(command_str)
+            if command_str is not None and ssh is not None:
+                stdin, stdout, stderr = ssh.exec_command(command_str)
+                res, err = stdout.read(), stderr.read()
+                re = res if res else err
+                result = re.decode()
+        except paramiko.AuthenticationException as ae:
+            LOG.error('doexec Authentication error:{}'.format(ae))
+            raise exception.InvalidUsernameOrPassword()
+        except Exception as e:
+            err = six.text_type(e)
+            LOG.error('doexec InvalidUsernameOrPassword error')
+            if 'timed out' in err:
+                raise exception.SSHConnectTimeout()
+            elif 'No authentication methods available' in err \
+                    or 'Authentication failed' in err:
+                raise exception.SSHInvalidUsernameOrPassword()
+            elif 'not a valid RSA private key file' in err:
+                raise exception.InvalidPrivateKey()
+            else:
+                raise exception.SSHException(err)
+        return result
+
+    def exec_ssh_command(self, command):
+        try:
+            with self.ssh_pool.item() as ssh:
+                ssh_info = SSHHandler.do_exec(command, ssh)
+            return ssh_info
+        except Exception as e:
+            msg = "Failed to ssh ibm storwize_svc %s: %s" % \
+                  (command, six.text_type(e))
+            raise exception.SSHException(msg)
+
+    def change_capacity_to_bytes(self, unit):
+        unit = unit.upper()
+        if unit == 'TB':
+            result = units.Ti
+        elif unit == 'GB':
+            result = units.Gi
+        elif unit == 'MB':
+            result = units.Mi
+        elif unit == 'KB':
+            result = units.Ki
+        else:
+            result = 1
+        return int(result)
+
+    def parse_string(self, value):
+        capacity = 0
+        if value:
+            if value.isdigit():
+                capacity = float(value)
+            else:
+                unit = value[-2:]
+                capacity = float(value[:-2]) * int(
+                    self.change_capacity_to_bytes(unit))
+        return capacity
+
+    def get_storage(self):
+        try:
+            system_info = self.exec_ssh_command('lssystem')
+            enclosure_info = self.exec_ssh_command('lsenclosure -delim :')
```

Comment:
Any command on the enclosure my take time ..Do we have any timeouts?

Reply:
Also, any command if hung, we should have a proper cleaup mechanism. Is that handled?